Bluetooth: Controller: Refactor same peer connection check and add check in Initiator

cvinayak opened the pull request:
Store and check device identity addr type of public or
random.

Add implementation to initiator to check and reject
connection requests to already connected peer.

These changes continue to pass the related 1 peripheral conformance test cases,
and now pass 3 initiator conformance tests.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Use id addr type to check same peer connection

Store and check device identity addr type of public or
random.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Add same peer connection check to initiator

Add implementation to initiator to check and reject
connection requests to already connected peer.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_filter.c:
```diff
@@ -1103,12 +1103,13 @@ static void conn_rpa_update(uint8_t rl_idx)
-		    conn->peer_addr_type < 2U &&
-		    !memcmp(conn->peer_addr, rl[rl_idx].curr_rpa.val, BDADDR_SIZE)) {
-			memcpy(conn->peer_addr, rl[rl_idx].id_addr.val, BDADDR_SIZE);
-			conn->peer_addr_type += 2U;
```
@wopu-ot What is the reason to store Public or Random ID used for resolvable private address? 1 bit device address type of public or random should suffice for same peer checks, right?

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_master.c:
```diff
@@ -93,6 +93,18 @@ uint8_t ll_create_connection(uint16_t scan_interval, uint16_t scan_window,
 		return BT_HCI_ERR_CMD_DISALLOWED;
 	}
 
+#if defined(CONFIG_BT_CTLR_CHECK_SAME_PEER_CONN)
+	uint8_t own_id_addr_type = (own_addr_type & 0x01);
+	uint8_t *own_id_addr;
+
+	/* Do not connect twice to the same peer */
+	own_id_addr = ll_addr_get(own_id_addr_type, NULL);
+	if (ull_conn_peer_connected(own_id_addr_type, own_id_addr,
+				    peer_addr_type, peer_addr)) {
+		return BT_HCI_ERR_CONN_ALREADY_EXISTS;
+	}
+#endif /* CONFIG_BT_CTLR_CHECK_SAME_PEER_CONN */
+
```
@wopu-ot I should resolved the supplied `peer_addr` and store the `peer_id_addr` instead, right?

cvinayak reviewed (commented):
@wopu-ot Need some clarifications on the privacy scenarios... see inline comments

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_master.c:
```diff
@@ -93,6 +93,18 @@ uint8_t ll_create_connection(uint16_t scan_interval, uint16_t scan_window,
 	}
 
+#if defined(CONFIG_BT_CTLR_CHECK_SAME_PEER_CONN)
+	uint8_t own_id_addr_type = (own_addr_type & 0x01);
+	uint8_t *own_id_addr;
```
```suggestion
	const uint8_t own_id_addr_type = (own_addr_type & 0x01);
	const uint8_t *own_id_addr;
```

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_slave.c:
```diff
@@ -112,22 +112,23 @@ void ull_slave_setup(struct node_rx_hdr *rx, struct node_rx_ftr *ftr,
 #if defined(CONFIG_BT_CTLR_CHECK_SAME_PEER_CONN)
-	uint8_t own_addr_type = pdu_adv->rx_addr;
-	uint8_t *own_addr = adv->own_addr;
+	uint8_t peer_id_addr_type = (peer_addr_type & 0x01);
+	uint8_t own_id_addr_type = pdu_adv->rx_addr;
+	uint8_t *own_id_addr = adv->own_id_addr;
```
```suggestion
	const uint8_t peer_id_addr_type = (peer_addr_type & 0x01);
	const uint8_t own_id_addr_type = pdu_adv->rx_addr;
	const uint8_t *own_id_addr = adv->own_id_addr;
```

Thalley reviewed (approved):
LGTM, but looks like we could add `const` a few more places after the function header was updated

cvinayak commented:
Added DNM to get review from @wopu-ot

cvinayak commented:
> 
> 
> Added DNM to get review from @wopu-ot

Can yourself @asmk-ot or @mtpr-ot review this one?

mtpr-ot reviewed (approved):
Looks good. Added @erbr-ot to confirm the id address changes.

cvinayak committed:
Bluetooth: Controller: Use define for invalid LLL context handle

Use a define for invalid LLL context handle instead of
0xFFFF magic value.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix missing reset of connection handle

Fix missing reset of connection handle in the LLL context.
During a central connection, if LL reset is called, then
disabled connection context's handle has to be reset
otherwise new connection creation will fail by detecting
that there exists a connection to same peer.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

erbr-ot reviewed (approved):
I tested this solution on our 2.2.1 derivative and it passes our complete bsim test suite which includes the EDTT HCI/GATT/LL tests
